File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int res = 0;
        dfs(root, res);
        return res;
    }

    int dfs(TreeNode* cur, int &res){
        if(!cur){
            return 0;
        }

        int left = dfs(cur->left,res);
        int right = dfs(cur->right,res);

        res = max( res, left + right);

        return 1 + max(left, right);
    }
};

File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.dfs(root, 1001, -1001)
        
    def dfs( self, root, upperbound, lowerbound):
        if not root: 
            return True
        elif root.val <= lowerbound or root.val >= upperbound:
            return False
        else: 
            return self.dfs(root.left, root.val, lowerbound) and self.dfs(root.right, upperbound, root.val)
